x-pack/platform/plugins/shared/observability_ai_assistant/server/utils/rewrite_query.ts

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { findLast, pick } from 'lodash';
+import { BoundInferenceClient } from '@kbn/inference-plugin/server';
+import { Message, MessageRole } from '../../common';
+import { KnowledgeBaseQueryContainer } from '../service/knowledge_base_service';
+
+export async function rewriteQuery({
+  inferenceClient,
+  messages,
+  systemMessage,
+  context,
+}: {
+  inferenceClient: BoundInferenceClient;
+  messages: Message[];
+  systemMessage?: string;
+  context: string;
+}): Promise<{
+  queries: KnowledgeBaseQueryContainer[];
+}> {
+  const lastUserMessage = serialize(
+    findLast(
+      messages,
+      (message) => message.message.role === MessageRole.User && !!message.message.name
+    )
+  );
+
+  function serialize(message?: Message) {
+    return message
+      ? JSON.stringify(pick(message.message, 'content', 'name', 'function_call', 'role'))
+      : '';
+  }
+
+  const serializedConversation = messages.map((message) => {
+    return JSON.stringify(pick(message.message, 'content', 'name', 'function_call', 'role'));
```

Comment:
Can you use `serialize` here?
```suggestion
    return serialize(message);
```